ssl/s3_enc.c

```diff
@@ -399,11 +418,14 @@ int ssl3_change_cipher_state(SSL *s, int which)
 
     OPENSSL_cleanse(&(exp_key[0]), sizeof(exp_key));
     OPENSSL_cleanse(&(exp_iv[0]), sizeof(exp_iv));
-    EVP_MD_CTX_cleanup(&md);
     return (1);
  err:
     SSLerr(SSL_F_SSL3_CHANGE_CIPHER_STATE, ERR_R_MALLOC_FAILURE);
  err2:
+    if (s->enc_write_ctx && !(which & SSL3_CC_READ)) {
+        EVP_CIPHER_CTX_free(s->enc_write_ctx);
+        s->enc_write_ctx = NULL;
+    }
```

Comment:
Don't you have the same problem with the enc_read_ctx? The code is at least the same, so it now doesn't seem to be consistent anymore.

I'm also not sure why it's needed that you do this. I understand that it's when you call do_ssl3_write() via ssl3_shutdown(). But it seems wrong to me that shutdown is called in this case, it shouldn't attempt to send anymore when it got a failure here. And I guess the code just doesn't know that this was a fatal error and shouldn't use that SSL anymore.